suseconnect/migration.go

```diff
@@ -0,0 +1,562 @@
+		//             (options[:non_interactive] ? "--non-interactive " : "") +
+		//             (options[:verbose] ? "--verbose " : "") +
+		//             (options[:quiet] ? "--quiet " : "") +
+		//             "--no-refresh patch --updatestack-only"
+		//       print "\nExecuting '#{cmd}'\n\n" unless options[:quiet]
+		//       system cmd
+
+		//       # stop infinite restarting
+		//       # check that the patches were really installed
+		//       cmd = "zypper " +
+		//         (options[:root] ? "--root #{options[:root]} " : "") +
+		//         "--non-interactive --quiet --no-refresh patch-check --updatestack-only > /dev/null"
+		//       if ! system cmd
+		//         print "patch failed, exiting.\n"
+		//         exit 1
+		//       end
+
+		//       print "\nRestarting the migration script...\n" unless options[:quiet]
+		//       exec $0, *save_argv
+		//     end
+		//     exit 1
+		//   end
+	}
+	connect.QuietOut.Print("\n")
+
+	// This is only necessary, if we run with --root option
+	if err := connect.RefreshRepos("", false, quiet, verbose, nonInteractive); err != nil {
+		fmt.Println("repository refresh failed, exiting")
+		os.Exit(1)
+	}
+
+	systemProducts, err := checkSystemProducts(true, true)
+	if err != nil {
+		fmt.Printf("Can't determine the list of installed products: %v\n", err)
+		os.Exit(1)
+	}
+
+	if len(systemProducts) == 0 {
+		fmt.Println("No products found, migration is not possible.")
+		os.Exit(1)
+	}
+
+	// This is not fully correct name as installedIDs could also contain some products
+	// which are activated but not installed. This matches the original implementation.
+	installedIDs := make(map[string]struct{}, 0)
+	for _, prod := range systemProducts {
+		installedIDs[prod.ToTriplet()] = struct{}{}
+	}
+
+	allMigrations := make([]connect.MigrationPath, 0)
+	// if options[:to_product]
+	//   begin
+	//     identifier, version, arch = options[:to_product].split('/')
+	//     new_product = OpenStruct.new(
+	//                                  identifier: identifier,
+	//                                  version:   version,
+	//                                  arch:       arch
+	//                                  )
+	//     migrations_all = SUSE::Connect::YaST.system_offline_migrations(system_products, new_product)
+	//   rescue => e
+	//     print "Can't get available migrations from server: #{e.class}: #{e.message}\n"
+	//     exit 1
+	//   end
+	// else
+	{
+		allMigrations, err = connect.ProductMigrations(systemProducts)
+		if err != nil {
+			fmt.Printf("Can't get available migrations from server: %v\n", err)
+			os.Exit(1)
+		}
+	}
+
+	// preprocess the migrations lists
+	migrations := make([]connect.MigrationPath, 0)
+	unavailableMigrations := make([]connect.MigrationPath, 0)
+	for _, m := range allMigrations {
+		mAvailable := true
+		for _, p := range m {
+			mAvailable = mAvailable && p.Available
+		}
+		// sort migrations to put already installed products last and base products first
```

Comment:
could the sorting be done in function with some unit tests?

Reply:
done.